base/reflection.jl

```diff
@@ -2540,8 +2540,22 @@ min_world(m::Core.CodeInstance) = m.min_world
 max_world(m::Core.CodeInstance) = m.max_world
 min_world(m::Core.CodeInfo) = m.min_world
 max_world(m::Core.CodeInfo) = m.max_world
+
+"""
+    get_world_counter()
+
+Returns the current maximum world-age counter. This counter is global and monotonically
+increasing.
+"""
 get_world_counter() = ccall(:jl_get_world_counter, UInt, ())
```

Comment:
You also made me realize we need to go through the users of this function (including those in src) and make sure they can guard against this value being injected. Most of them should already ignore updates that try to use this value, as they would result in mis-compilations.

But we also need to update `jl_type_infer` to permit those updates, specifically for inference (since inference is special in that it knows how to deal with and avoid those miscompilations, specifically because it is the owner of those caches)

```diff --git a/src/gf.c b/src/gf.c
index efe624f518..e8a5f7450b 100644
--- a/src/gf.c
+++ b/src/gf.c
@@ -373,6 +373,8 @@ jl_code_info_t *jl_type_infer(jl_method_instance_t *mi, size_t world, int force)
 #ifdef _OS_WINDOWS_
     DWORD last_error = GetLastError();
 #endif
+    int last_pure = ct->ptls->in_pure_callback;
+    ct->ptls->in_pure_callback = 0;
     size_t last_age = ct->world_age;
     ct->world_age = jl_typeinf_world;
     mi->inInference = 1;
@@ -409,6 +411,7 @@ jl_code_info_t *jl_type_infer(jl_method_instance_t *mi, size_t world, int force)
     }
     ct->world_age = last_age;
     ct->reentrant_timing -= 0b10;
+    ct->ptls->in_pure_callback = last_pure;
     mi->inInference = 0;
 #ifdef _OS_WINDOWS_
     SetLastError(last_error);
```